airbyte-featureflag/src/main/kotlin/Client.kt

```diff
@@ -0,0 +1,196 @@
+        }
+    }
+}
+
+/**
+ * Test feature-flag client. Intended only for usage in testing scenarios.
+ *
+ * All [Flag] instances will use the provided [values] map as their source of truth, including [EnvVar] flags.
+ *
+ * @param [values] is a map of [Flag.key] to enabled/disabled status.
+ */
+class TestClient(val values: Map<String, Boolean>) : FeatureFlagClient {
+    override fun enabled(flag: Flag, ctx: Context): Boolean {
+        return when (flag) {
+            is EnvVar -> {
+                // convert to a EnvVar flag with a custom fetcher that uses the [values] of this Test class
+                // instead of fetching from the environment variables
+                EnvVar(envVar = flag.key, default = flag.default, team = flag.team) {
+                    values[flag.key]?.toString() ?: flag.default.toString()
+                }.enabled()
+            }
+
+            else -> values[flag.key] ?: flag.default
+        }
+    }
+}
+
+
+/**
+ * Data wrapper around OSS feature-flag configuration file.
+ *
+ * The file has the format of:
+ * flags:
+ *  - name: feature-one
+ *    enabled: true
+ *  - name: feature-two
+ *    enabled: false
+ */
+private data class PlatformFlags(val flags: List<PlatformFlag>)
+
+/**
+ * Data wrapper around an individual flag read from the configuration file.
+ */
+private data class PlatformFlag(val name: String, val enabled: Boolean)
+
+
+/** The yaml mapper is used for reading the feature-flag configuration file. */
+private val yamlMapper = ObjectMapper(YAMLFactory()).registerKotlinModule()
+
+/**
+ * Reads a yaml configuration file, converting it into a map of flag name to flag configuration.
+ *
+ * @param [path] to yaml config file
+ * @return map of feature-flag name to feature-flag config
+ */
+//private fun readConfig(path: Path): Map<String, OSSFlag> = yamlMapper.readValue<List<OSSFlag>>(path.toFile())
+//    .associateBy { it.name }
+private fun readConfig(path: Path): Map<String, PlatformFlag> = yamlMapper.readValue<PlatformFlags>(path.toFile())
+    .let { it.flags }
+    .associateBy { it.name }
+
+/**
+ * Monitors a [Path] for changes, calling [block] when a change is detected.
+ *
+ * @receiver Path
+ * @param [block] function called anytime a change is detected on this [Path]
+ */
+private fun Path.onChange(block: () -> Unit) {
+    val watcher = fileSystem.newWatchService()
+    // The watcher services requires a directory to be provided, hence the `parent` reference here as `this` path should be a file.
+    parent.register(watcher, StandardWatchEventKinds.ENTRY_MODIFY, StandardWatchEventKinds.ENTRY_CREATE)
+
+    thread(isDaemon = true, name = "feature-flag-watcher", priority = MIN_PRIORITY) {
+        val key = watcher.take()
+        // The context on the poll-events for ENTRY_MODIFY events should not be null, this is a sanity check more than anything else
+        key.pollEvents().mapNotNull { it.context() as? Path }
```

Comment:
Nit/QQ about kotlin, is the `?` needed in the `as?` since we know that the map only map the non null value.
Also do we need to filter out null values or is it included in the `mapNonNull` function?

Reply:
> is the ? needed in the as?

As Kotlin requires null values to be explicitly marked as nullable, this also applies when casting one class to another.
If `context()` returned a null (which it can according to the [javadocs](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/nio/file/WatchEvent.html#context())) then the following code would result in an exception when `context()` was `null`
```kotlin
context() as Path
```

Kotlin as a ["Safe" (nullable) cast operator](https://kotlinlang.org/docs/typecasts.html#safe-nullable-cast-operator) (rolls right off the tonguq) for handling this scenario.  If `context()` returns `null`, the result will be `null` instead of an exception. If `context()` is not-null then it will be cast to a `Path` (which could also throw an exception if the type isn't a `Path`, but according to the javadocs, it should be a `Path`).

> do we need to filter out null values or is it included in the mapNonNull function

The [mapNotNull(https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/map-not-null.html) filters out null values. I had originally written this as two separate methods (one for the map and one for filtering out nulls) and intellij told me to collapse them into this one method call.